safe-transaction-client/src/lib.rs

```diff
@@ -0,0 +1,343 @@
+            nonce: U256::from(nonce),
+            operation,
+        })
+    }
+
+    pub fn propose(&self, signed_safe_tx: SignedSafeTx) -> Result {
+        let SafeTx {
+            to,
+            value,
+            data,
+            nonce,
+            operation,
+            ..
+        } = signed_safe_tx.inner;
+        self.client
+            .agent
+            .post(&format!(
+                "{}/v1/safes/{}/multisig-transactions/",
+                self.client.transactions_api,
+                to_checksum(&self.safe_address, None),
+            ))
+            .send_json(ureq::json!({
+                "to": to_checksum(&to, None),
+                "value": format!("{}", value),
+                "data": data,
+                "operation": operation as u8,
+                "gasToken": None::<()>,
+                "safeTxGas": "0",
+                "baseGas": "0",
+                "gasPrice": "0",
+                "refundReceiver": Address::zero(),
+                "nonce": format!("{}", nonce),
+                "contractTransactionHash": signed_safe_tx.safe_tx_hash,
+                "sender": to_checksum(&signed_safe_tx.sender, None),
+                "signature": format!("0x{}", signed_safe_tx.signature),
+                "origin": "An unauthenticated origin string?",
+            }))
+            // 200 OK, is not okay. We expect 201 Created.
+            .and_then(|r| match r.status() {
+                201 => Ok(()),
+                s => Err(ureq::Error::Status(s, r)),
+            })
+            .map_err(From::from)
+    }
+
+    pub fn add_to_confirmations(&self, signed_safe_tx: SignedSafeTxHash) -> Result {
+        self.client
+            .agent
+            .post(&format!(
+                "{}/v1/multisig-transactions/{:?}/confirmations/",
+                self.client.transactions_api, signed_safe_tx.safe_tx_hash,
+            ))
+            .send_json(ureq::json!({
+                "signature": format!("0x{}", signed_safe_tx.signature),
+            }))
+            // 200 OK, is not okay. We expect 201 Created.
+            .and_then(|r| match r.status() {
+                201 => Ok(()),
+                s => Err(ureq::Error::Status(s, r)),
+            })
+            .map_err(From::from)
+    }
+}
+
+impl SafeTx {
+    // We consume self. If signing fails, we have bigger issues than recreating the transaction.
+    pub async fn sign<S>(self, signer: &S) -> Result<SignedSafeTx, Error<S::Error>>
+    where
+        S: Signer,
+        Error<S::Error>: From<S::Error>,
+    {
+        use ethers::abi::Tokenizable;
+        use ethers::utils::keccak256;
+        use tiny_keccak::{Hasher, Keccak};
+        macro_rules! tokenize {
+            ($($tok:expr),* $(,)*) => {
+                &[$($tok.into_token()),*][..]
+            }
+        }
+
+        let safe_tx_hash = TxHash::from({
+            let mut hasher = Keccak::v256();
+            let mut output = [0u8; 32];
+            hasher.update(&[0x19, 0x01]);
```

Comment:
What's this again?

Reply:
The magic bytes? Ask the Gnosis devs.

https://github.com/gnosis/safe-contracts/blob/main/contracts/GnosisSafe.sol#L393

The multi-update hasher is just `keccak256(abi.encodePacked(..))` where the latter function doesn't exist in `ethers::abi` and would just create a vector just to read it back anyway.

Reply:
Ah it's `eth_sign`: https://eips.ethereum.org/EIPS/eip-191